src/exo/LoopIR_scheduling.py

```diff
@@ -3388,30 +3513,37 @@ def __init__(self, proc, alloc_site, bounds):
         self.bounds = bounds
         super().__init__(proc)
 
-    def map_s(self, s):
-        if s is self.alloc_site:
-            assert isinstance(s.type, T.Tensor)
-            if len(self.bounds) != len(s.type.hi):
-                raise SchedulingError(
-                    f'bound_alloc: dimensions do not match: {len(self.bounds)} '
-                    f'!= {len(s.type.hi)} (expected)')
-
-            new_type = T.Tensor(
-                [(new if new else old)
-                 for old, new in zip(s.type.hi, self.bounds)],
-                s.type.is_window,
-                s.type.type,
-            )
+    def map_stmts(self, stmts):
+        new_stmts = []
+        for i,s in enumerate(stmts):
+            if s is self.alloc_site:
+                assert isinstance(s.type, T.Tensor)
+                if len(self.bounds) != len(s.type.hi):
+                    raise SchedulingError(
+                        f'bound_alloc: dimensions do not match: '
+                        f'{len(self.bounds)} != {len(s.type.hi)} (expected)')
 
-            return [LoopIR.Alloc(s.name, new_type, s.mem, s.eff, s.srcinfo)]
+                new_bounds = [ new if new else old
+                               for old, new in zip(s.type.hi, self.bounds) ]
+                newtyp = T.Tensor(new_bounds, s.type.is_window, s.type.type)
 
-        return super().map_s(s)
+                # TODO: CHECK THE BOUNDS OF ACCESSES IN stmts[i+1:] here
+
+                s = LoopIR.Alloc(s.name, newtyp, s.mem, s.eff, s.srcinfo)
+                return new_stmts + [s] + stmts[i+1:]
+            else:
+                new_stmts += self.map_s(s)
+
+        return new_stmts
 
 # --------------------------------------------------------------------------- #
 # --------------------------------------------------------------------------- #
 # The Passes to export
 
 class Schedules:
+    DoRenameBuf = _DoRenameBuf
+    DoAssignNewVarBefore = _DoAssignNewVarBefore
+    DoDeleteDeadCode = _DoDeleteDeadCode
```

Comment:
We should go through the motivation of these primitives together and also write tests